src/provider/model/Model.ts

```diff
@@ -252,121 +267,135 @@ export class Model {
+        this._connectionsById[connection.id] = connection;
+        delete this._expectedConnectionsById[connection.id];
 
-        console.info(`Registering window ${appWindow.id}`);
-        this._windowsById[appWindow.id] = appWindow;
-        delete this._expectedWindowsById[id];
+        this.onConnectionAdded.emit(connection);
+        this._onConnectionRegisteredInternal.emit();
+    }
 
-        this.onWindowAdded.emit(appWindow);
-        this._onWindowRegisteredInternal.emit();
+    private removeConnection(identity: Identity): void {
+        const id: string = getId(identity);
+        const connection: AppConnection = this._connectionsById[id];
 
-        return appWindow;
+        if (connection) {
+            delete this._connectionsById[id];
+            this.onConnectionRemoved.emit(connection);
+        } else if (this._expectedConnectionsById[id]) {
+            delete this._expectedConnectionsById[id];
+        }
     }
 
-    private findWindowsByAppId(appId: AppId): AppWindow[] {
-        return this.findWindows(appWindow => appWindow.appInfo.appId === appId);
+    private findConnectionsByAppId(appId: AppId): AppConnection[] {
+        return this.findConnections(connection => connection.appInfo.appId === appId);
     }
 
-    private findWindows(predicate: (appWindow: AppWindow) => boolean): AppWindow[] {
-        return this.windows.filter(predicate);
+    private findConnections(predicate: (connection: AppConnection) => boolean): AppConnection[] {
+        return this.connections.filter(predicate);
     }
 
-    private getOrCreateExpectedWindow(identity: Identity): ExpectedWindow {
+    private getOrCreateExpectedConnection(identity: Identity): ExpectedConnection {
         const id = getId(identity);
 
-        if (this._expectedWindowsById[id]) {
-            return this._expectedWindowsById[id];
+        if (this._expectedConnectionsById[id]) {
+            return this._expectedConnectionsById[id];
         } else {
             // A promise that never resolves but rejects when the window has closed
-            const closed = allowReject(untilSignal(this._environment.windowClosed, (testIdentity) => {
+            const closed = allowReject(untilSignal(this._environment.onWindowClosed, (testIdentity) => {
                 return getId(testIdentity) === id;
             }).then(() => {
                 throw new Error(EXPECT_CLOSED_MESSAGE);
             }));
 
             // Create a promise that resolves once the window has been created
-            const created = untilTrue(this._environment.windowCreated, () => {
-                return this._environment.isWindowCreated(identity);
+            const windowCreated = untilTrue(this._environment.onWindowCreated, () => {
+                return this._environment.isKnownEntity(identity);
+            });
+            const connectionCreated = untilTrue(this.onConnectionAdded, () => {
+                return this._apiHandler.isClientConnection(identity);
```

Comment:
This `this._apiHandler.onConnection` would be the correct signal to use here

Reply:
Was using this before, but was causing issues due to the delay between `ApiHandler.onConnection` and `Model.onConnectionAdded` - since the handler in between those signals has async stuff in it.

Have switched back to the ApiHandler signal and all seems fine now though.